web/src/components/organisms/Project/ModelsMenu/hooks.ts

```diff
@@ -0,0 +1,117 @@
+import { useCallback, useEffect, useMemo, useState } from "react";
+
+import { Model } from "@reearth-cms/components/molecules/Schema/types";
+import {
+  useGetModelsQuery,
+  useCreateModelMutation,
+  useCheckModelKeyAvailabilityLazyQuery,
+  Model as GQLModel,
+} from "@reearth-cms/gql/graphql-client-api";
+
+type Params = {
+  projectId?: string;
+  modelId?: string;
+};
+
+export default ({ projectId, modelId }: Params) => {
+  const [modelModalShown, setModelModalShown] = useState(false);
+  const [isKeyAvailable, setIsKeyAvailable] = useState(false);
+  const [CheckModelKeyAvailability, { data: keyData }] = useCheckModelKeyAvailabilityLazyQuery({
+    fetchPolicy: "no-cache",
+  });
+
+  const handleModelKeyCheck = useCallback(
+    async (projectId: string, key: string) => {
+      if (!projectId || !key) return false;
+      const response = await CheckModelKeyAvailability({ variables: { projectId, key } });
+      return response.data ? response.data.checkModelKeyAvailability.available : false;
+    },
+    [CheckModelKeyAvailability],
+  );
+
+  useEffect(() => {
+    setIsKeyAvailable(keyData?.checkModelKeyAvailability.available ?? false);
+  }, [keyData?.checkModelKeyAvailability]);
+
+  const { data } = useGetModelsQuery({
+    variables: { projectId: projectId ?? "", first: 100 },
+    skip: !projectId,
+  });
+
+  const fromModel = (model: GQLModel) => ({
+    id: model.id,
+    description: model.description,
+    name: model.name,
+    key: model.key,
+    schema: {
+      id: model.schema?.id,
+      fields: model.schema?.fields.map(field => ({
+        id: field.id,
+        description: field.description,
+        title: field.title,
+        type: field.type,
+        key: field.key,
+        unique: field.unique,
+        required: field.required,
+        typeProperty: field.typeProperty,
+      })),
+    },
+  });
+
+  const models = useMemo(() => {
+    return (data?.models.nodes ?? [])
+      .map<Model | undefined>(model => (model ? fromModel(model as GQLModel) : undefined))
+      .filter((model): model is Model => !!model);
+  }, [data?.models.nodes]);
+
+  const rawModel = useMemo<GQLModel | undefined>(
+    () => (data?.models.nodes as GQLModel[])?.find((node: GQLModel) => node.id === modelId),
```

Comment:
Were you getting an error without doing `as GQLModel[]`? I feel like `data.models.nodes` should be already typed as the GraphQL Model type. 
I'm sure you already know, but you should try all other possibilities before typecasting. Typecasting should be a last last last resort

Reply:
If you can do without the typecasting, please update everywhere you are using typecasting

Reply:
Please advise, I couldn't find another way.
If I remove the casting it shows an error.
![2](https://user-images.githubusercontent.com/78056580/193777647-8f0af6be-bcea-4e0b-9823-fa18d540af47.png)

Reply:
That is because `data?.models/nodes?` might be undefined/null and you are trying to assign it as only `GQLModel`. 
Can you try typing it as `GQLModel | null` or `GQLModel | null | undefined` and see if it is solved?
If not, try removing the type completely and see if you get any errors on the return type being wrong.